### PR TITLE
Support deploying to gh-pages via CircleCI

### PR DESCRIPTION
**Pull request by raysonkoh**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] Others, please explain:

Related to #1432 

**Overview of changes:**
Add logic to detect CircleCI environment using `$CIRCLECI` and extract the reposlug using the combination of `$CIRCLE_PROJECT_USERNAME/$CIRCLE_PROJECT_REPONAME`.

**Anything you'd like to highlight / discuss:**
- I have added `skip ci` to the commit message when pushing to the `gh-pages` branch as Circle CI would not ignore the `gh-pages` branch. Upon further research, it seems as though other CI platforms are respecting the `skip ci` directive also. Would there be any concerns with adding `skip ci` to the commit message?
Ref:
  - https://circleci.com/docs/2.0/skip-build/#skipping-a-build
  - https://docs.travis-ci.com/user/customizing-the-build/#skipping-a-build
  - https://www.appveyor.com/docs/how-to/filtering-commits/#skip-directive-in-commit-message
  - Github Action doesn't seem to support `skip ci` :disappointed:. There is an open issue in actions/runner#774 **Edit: Github Actions started supporting `skip ci` as of 8 Feb 2021:** https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/

**Testing instructions:**
1. Sign up on [CircleCI](https://circleci.com/) using your Github account. 
2. In the projects dashboard, click on `Set Up Project` on the repo of your markbind site.
3. Create a [Github personal access token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) with **repo** permissions and add the generated personal access token as a custom environmental variable in "Project Settings" -> "Environment Variables"
4. In your the markbind site repo, commit and push the following content into `PROJECT_ROOT/.circleci/config.yml`:
```
jobs:
  Build-And-Deploy:
    docker:
      - image: 'cimg/base:stable'
    steps:
      - checkout
      - node/install:
          node-version: "10"
          npm-version: "6"
          install-yarn: false
      - run: node --version
      - run: cd ../ && git clone -b fix-1432-circle-ci-1 https://github.com/raysonkoh/markbind.git
      - run: cd ../markbind/packages/cli && npm link
      - run: cd ../markbind && npm run setup
      - run: markbind build
      - run: markbind deploy --ci
version: 2.1
orbs:
  node: circleci/node@4.1.0
workflows:
  Deploy-Markbind-Site:
    jobs:
      - Build-And-Deploy
```

**Proposed commit message: (wrap lines at 72 characters)**
Support deploying to gh-pages via CircleCI

As the markbind deploy --ci command would
push a commit to the gh-pages repo, it might 
unintentionally trigger a CI build.

Let's insert a [skip ci] directive into the commit 
message, which instructs the CI services to ignore
that commit and thus prevent it from unintentionally
triggering a CI build.

---

**Checklist:** :ballot_box_with_check:

- [x] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No blatantly unrelated changes
- [x] Pinged someone for a review!

### CODE REVIEW

File: packages/core/src/Site/index.js [hunk 1]
```diff
@@ -1153,38 +1154,36 @@ class Site {
       let repoSlug;
 
       if (process.env.TRAVIS) {
-        if (options.repo) {
-          repoSlug = Site.extractRepoSlug(options.repo);
-        } else {
-          repoSlug = process.env.TRAVIS_REPO_SLUG;
-        }
+        repoSlug = Site.extractRepoSlug(options.repo, process.env.TRAVIS_REPO_SLUG);
```
Comment (wxwxwxwx9) [inline packages/core/src/Site/index.js:1157]: Nice workaround! 👍

**Comment by wxwxwxwx9** [comment]
Hi @raysonkoh, good job on this PR :-) I've tested the deployment on CircleCI and it works on my side. 

I think that the user guide was updated pretty well. The instructions were clear with suitable screenshots at right places. I also like how you shifted the token generation sections to the top and referred the user back to the appropriate ones in the respective CI section (e.g. refer users to generating personal access token in CircleCI). 

Since I don't have much experience in CI, I've generally referred to your PR #1453 and #1460 as a guideline of how this PR should look like, and I think generally it looks fine to me. 

But there's the part about `skip-ci` which I'm not entirely sure about. Can I get your help to provide more context on it?

File: docs/userGuide/deployingTheSite.md [hunk 2]
```diff
@@ -137,8 +145,6 @@ Since May 2018, Travis CI has been [undergoing migration to `travis-ci.com`](htt
 
 ##### Configuring your repository in Travis CI
 
-1. [Generate a GitHub personal access token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/#creating-a-token) with **repo** permissions. Take note of the generated token - you will not be able to see it again once you navigate away from the page.
-<include src="screenshot.md" boilerplate var-alt="GitHub Token Repo Permissions" var-file="githubTokenRepoPermissions.png" inline />
 1. [Add an environment variable in Travis CI](https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings) named `GITHUB_TOKEN`, with the value set to the personal access token generated in the previous step. ==Ensure that _Display value in the build log_ is set to _Off_.==
```
Comment (wxwxwxwx9) [inline docs/userGuide/deployingTheSite.md:148]: Regarding the "previous step", I think we should link it back to the *personal access token section* above.

Comment (raysonkoh) [reply docs/userGuide/deployingTheSite.md]: Thanks for the catch @wxwxwxwx9 !

**Comment by raysonkoh** [comment]
> But there's the part about skip-ci which I'm not entirely sure about. Can I get your help to provide more context on it?

So when we do `markbind deploy --ci`, it will build the markbind site and push the changes to the `gh-pages` branch of the repo of the markbind site. 

In Circle CI, the act of pushing to the `gh-pages` branch would trigger a CI build. That CI build would subsequently fail because there is no `config.yml` file in the `gh-pages` branch.

So the purpose of `skip ci` in the commit message is that it is a directive that explicitly instructs CI platforms to ignore that specific `"Site Update. [skip ci]"` commit to the `gh-pages` branch which may unintentionally trigger a CI build.

**Comment by wxwxwxwx9** [comment]
> > But there's the part about skip-ci which I'm not entirely sure about. Can I get your help to provide more context on it?
> 
> So when we do `markbind deploy --ci`, it will build the markbind site and push the changes to the `gh-pages` branch of the repo of the markbind site.
> 
> In Circle CI, the act of pushing to the `gh-pages` branch would trigger a CI build. That CI build would subsequently fail because there is no `config.yml` file in the `gh-pages` branch.
> 
> So the purpose of `skip ci` in the commit message is that it is a directive that explicitly instructs CI platforms to ignore that specific `"Site Update. [skip ci]"` commit to the `gh-pages` branch which may unintentionally trigger a CI build.

Understood. Thanks for the explanation!

> Github Action doesn't seem to support skip ci 😞. There is an open issue in actions/runner#774

I noticed that GitHub Action has just started supporting `[skip ci]` a few days ago. Somebody brought it up in that open issue you linked. So I think using `[skip ci]` would be the right way to go, since it seems to be the convention for all the CI platforms we use so far :-)

https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/

Everything LGTM! Just need input from @ang-zeyu

File: packages/core/test/unit/Site.test.js [hunk 3]
```diff
@@ -345,14 +362,21 @@ describe('Site deploy with various CI environments', () => {
...
+    ['GITHUB_ACTIONS', { reposlug: 'GITHUB_REPOSITORY' }],
+    ['CIRCLECI', { username: 'CIRCLE_PROJECT_USERNAME', reponame: 'CIRCLE_PROJECT_REPONAME' }],
   ])('Site deploy -c/--ci deploys to correct repo when .git is in repo name',
      async (ciIdentifier, repoSlugIdentifier) => {
        process.env[ciIdentifier] = true;
-       process.env[repoSlugIdentifier] = 'GENERIC_USER/GENERIC_REPO.github.io';
        process.env.GITHUB_TOKEN = 'githubToken';
+       const genericRepoSlug = 'GENERIC_USER/GENERIC_REPO.github.io';
+       if (repoSlugIdentifier.reposlug) {
+         process.env[repoSlugIdentifier.reposlug] = 'GENERIC_USER/GENERIC_REPO.github.io';
```
Comment (ang-zeyu) [inline packages/core/test/unit/Site.test.js:375]: could use `genericRepoSlug` here too

**Review by ang-zeyu — COMMENTED** [review]
Thanks! @raysonkoh 

Just one nit, everything else looks good 🙂 

> I have added skip ci to the commit message when pushing to the gh-pages branch as Circle CI would not ignore the gh-pages branch. Upon further research, it seems as though other CI platforms are respecting the skip ci directive also. Would there be any concerns with adding skip ci to the commit message?

This might prevent adding further user workflows acting on the `gh-pages` branch, but as it is currently (all files are removed before pushing) I don't think it would be a breaking change since any CI config files would be removed

**Comment by raysonkoh** [comment]
Thanks @ang-zeyu for the review! I have made the change to the test.

**Review by ang-zeyu — APPROVED** [review]
Lgtm! 👍

**Pull request merged**